### PR TITLE
Complete notification

### PR DESCRIPTION
**Pull request by sergiwez**
**CHECK LIST**
- [x] specs added
- [x] specs passed
- [x] code coverage >90%
- [x] DB indexes added if needed
- [x] seed.rb file is updated if needed
- [x] PR is reviewed manually again (to make sure you have 100% ready code)
- [x] PR doesn't have conflicts with main branch
- [x] I've checked new feature as logged in and logged out user if needed
- [x] "let!" is used in specs
- [ ] screenshot of rails_best_practices and other tools added
- [x] screenshot of feature is added if task has UI

image

### CODE REVIEW

File: app/controllers/account/notifications_controller.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+class Account::NotificationsController < ApplicationController
+  def index; end
+
+  def read
+    @notification = resource
+    @notification.mark_as_readed
+    @notifications_count = collection.unread.count
+    render 'account/notifications/read'
```
Comment (romaslmd) [inline app/controllers/account/notifications_controller.rb]: redundant `render`

File: app/helpers/application_helper.rb [hunk 2]
```diff
@@ -19,6 +19,14 @@ def impersonated?
     current_user != true_user
   end
 
+  def notifications_count
+    notifications.count
```
Comment (romaslmd) [inline app/helpers/application_helper.rb]: why is this in helper?

File: app/libs/notification_builder.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+  def find_user(user)
+    if content.is_a?(Company)
+      @user = content.user
+    elsif content.is_a?(Page)
+      @user = content.company.user
+    elsif user.present?
+      @user = user
+    else
+      raise ArgumentError('Specify user')
```
Comment (romaslmd) [inline app/libs/notification_builder.rb]: `NotImplementedError` sounds better

File: app/libs/notification_builder.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
+class NotificationBuilder
+  attr_reader :content, :status, :notification_content, :user
+
+  def initialize(content: nil, user: nil)
+    raise ArgumentError('specify content') unless content
```
Comment (romaslmd) [inline app/libs/notification_builder.rb]: multiple validations used
refer `#find_user`

File: app/libs/notification_builder.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    @content = content
+    find_user(user)
```
Comment (romaslmd) [inline app/libs/notification_builder.rb]: mb `#set_user`

Comment (sergiwez) [reply app/libs/notification_builder.rb]: image
Robocop doesn't like it

File: app/helpers/application_helper.rb [hunk 6]
```diff
@@ -19,6 +19,10 @@ def impersonated?
     current_user != true_user
   end
 
+  def notifications
+    current_user.notifications.unread.order('created_at DESC')
```
Comment (igorkasyanchuk) [inline app/helpers/application_helper.rb]: .order('created_at DESC') make as scope

+ order(created_at: :desc)

File: app/libs/notification_builder.rb [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+  def find_user
+    if content.is_a?(Company)
+      @user = content.user
+    elsif content.is_a?(Page)
+      @user = content.company.user
+    end
+  end
+
+  def success(content)
```
Comment (igorkasyanchuk) [inline app/libs/notification_builder.rb]: content why you passing content? you already have content

File: app/models/company.rb [hunk 8]
```diff
@@ -35,6 +35,7 @@ class Company < ApplicationRecord
   belongs_to :user
   belongs_to :tenant, optional: true
   has_many   :pages, dependent: :destroy
+  has_many   :notifications, as: :notificable, dependent: :destroy
```
Comment (igorkasyanchuk) [inline app/models/company.rb]: dependent: :delete_all

File: app/workers/new_page_worker.rb [hunk 9]
```diff
@@ -17,6 +18,10 @@ def process
     page.update_attributes(status: Page::IN_PROGRESS_STATUS)
     parse_html_content
     make_screenshot
+    NotificationBuilder
+      .new(content: page)
+      .success("Page #{page.id} was successful processed")
```
Comment (igorkasyanchuk) [inline app/workers/new_page_worker.rb]: I see ... change param from `content` to message

File: db/migrate/20180203214857_create_notifications.rb [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+  def change
+    create_table :notifications do |t|
+      t.string  :status
+      t.boolean :readed, default: false
+      t.string  :content
+      t.integer :user_id
+      t.integer :notificable_id
+      t.string  :notificable_type
+      t.index :user_id
+      t.index :notificable_id
```
Comment (igorkasyanchuk) [inline db/migrate/20180203214857_create_notifications.rb]: bad index by _id, _type ask friends

**Pull request merged**